### PR TITLE
Add ElasticSearch 7.x support

### PR DESCRIPTION
**Pull request by harudark**
# Description
This pull request adds ES 7.x support in Bosun.
I submitted https://github.com/bosun-monitor/bosun/pull/2427 before but that is based on my own go modules migration pull request which is different from what upstream maintainers did. This one is based on upstream go modules migration.

- https://github.com/olivere/elastic/ version 7 client is used.
- Most code are copy-pasted from ES version 6 support.
- The one difference is in ES 7 [mapping types is removed](https://www.elastic.co/guide/en/elasticsearch/reference/master/removal-of-types.html), [Type(docType)](https://github.com/bosun-monitor/bosun/blob/f4392c2a8852b753e7490c27a599c04876c912c7/annotate/backend/elastic6.go#L161) are removed in ES7

## Type of change

From the following, please check the options that are relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How has this been tested?

Tested with ES 7.2 and also ES 6.8 to make sure nothing breaks.
- update the configuration toml file and create some alerts to query ES
- verify alerts can be triggered and no errors in ES query
- verify annotate can work

# Checklist:

- [x] This contribution follows the project's [code of conduct](https://github.com/bosun-monitor/bosun/blob/master/CODE_OF_CONDUCT.md)
- [x] This contribution follows the project's [contributing guidelines](https://github.com/bosun-monitor/bosun/blob/master/CONTRIBUTING.md)
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -33,22 +33,18 @@ require (
...
-	github.com/golang/protobuf v1.2.1-0.20190109072247-347cf4a86c1c // indirect
+	github.com/golang/groupcache v0.0.0-20190702054246-869f871628b6
 	github.com/gomodule/redigo v2.0.0+incompatible // indirect
 	github.com/google/cadvisor v0.21.1-0.20160212224345-e9739af18411
 	github.com/google/go-github v0.0.0-20151126072848-44b1ede22d71
 	github.com/google/go-querystring v0.0.0-20151028211038-2a60fc2ba6c1 // indirect
-	github.com/gorilla/context v1.1.1 // indirect
 	github.com/gorilla/css v0.0.0-20150317222238-a80e24ada269 // indirect
-	github.com/gorilla/mux v0.0.0-20160920230813-757bef944d0f
+	github.com/gorilla/mux v1.6.2
```
Comment (muffix) [inline go.mod]: Why has this dependency changed? `1.6.2` is almost two years newer than the previously pinned version and I can't see it being a direct dependency of any of the elastic modules? Similar for the other dependency changes - not saying that they're not worth bumping, just wondering why that happened here?

**Comment by svagner** [comment]
@harudark Should we close MRs https://github.com/bosun-monitor/bosun/pull/2426, https://github.com/bosun-monitor/bosun/pull/2427 and https://github.com/bosun-monitor/bosun/pull/2442 then?

File: annotate/backend/elastic7.go [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"reflect"
+	"strings"
+	"time"
+
+	"bosun.org/annotate"
+	elastic "github.com/olivere/elastic/v7"
```
[the remark this answers is not in the document]
Comment (harudark) [reply annotate/backend/elastic7.go]: You're right. I'll fix it.

File: annotate/backend/elastic7.go [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+)
+
+type Elastic7 struct {
```
[the remark this answers is not in the document]
Comment (harudark) [reply annotate/backend/elastic7.go]: Yeah, it makes sense to refactor it.
Also in bosun doc, it doesn't mention the supported ElasticSearch versions.
https://www.elastic.co/support/eol
Probably we can also remove ES 2.x and 5.x support and cleanup code.

File: go.sum [hunk 4]
```diff
@@ -37,10 +42,13 @@ github.com/alicebob/miniredis v2.5.0+incompatible h1:yBHoLpsyjupjz3NL3MhKMVkR41j
 github.com/alicebob/miniredis v2.5.0+incompatible/go.mod h1:8HZjEj4yU0dwhYHky+DxYx+6BMjkBbe5ONFIF1MXffk=
 github.com/andybalholm/cascadia v0.0.0-20150730174459-3ad29d1ad1c4 h1:I6VYmq9mpk/mxS0TSM46QHNZtkdHv9Tga/E8Pr7Tiac=
 github.com/andybalholm/cascadia v0.0.0-20150730174459-3ad29d1ad1c4/go.mod h1:3I+3V7B6gTBYfdpYgIG2ymALS9H+5VDKUl3lHH7ToM4=
+github.com/apache/thrift v0.12.0/go.mod h1:cp2SuWMxlEZw2r+iP2GNCdIi4C1qmUzdZFSVb+bacwQ=
 github.com/armon/go-metrics v0.0.0-20180917152333-f0300d1749da/go.mod h1:Q73ZrmVTwzkszR9V5SSuryQ31EELlFMUz1kKyl939pY=
 github.com/aws/aws-sdk-go v0.0.0-20180507225419-00862f899353/go.mod h1:ZRmQr0FajVIyZ4ZzBYKG5P3ZqPz9IHG41ZoMu1ADI3k=
 github.com/aws/aws-sdk-go v1.1.33 h1:AebOM+6ucC+bSl36BQrYkwShiqruxxJrZKduAvXSD1A=
 github.com/aws/aws-sdk-go v1.1.33/go.mod h1:ZRmQr0FajVIyZ4ZzBYKG5P3ZqPz9IHG41ZoMu1ADI3k=
+github.com/aws/aws-sdk-go v1.25.25 h1:j3HLOqcDWjNox1DyvJRs+kVQF42Ghtv6oL6cVBfXS3U=
```
[the remark this answers is not in the document]
Comment (harudark) [reply go.sum]: Sorry about that, I forgot it and will fix it.

File: go.mod (hunk 1 above)
Comment (harudark) [reply go.mod]: Thank you very much for your code review!

This dependency upgrade is unintended.
In ES 7 client go.mod(https://github.com/olivere/elastic/blob/v7.0.9/go.mod#L14-L15), it lists `go.opencensus.io v0.22.1` as direct dependency and `google.golang.org/api v0.3.1` as indirect dependency, which isn't right. I submitted an issue here(https://github.com/olivere/elastic/issues/1266).

`google.golang.org/api v0.3.1` also requires `go.opencensus.io`, but it requires an old version v0.20.1. https://github.com/googleapis/google-api-go-client/blob/v0.3.1/go.mod

For `go.opencensus.io` module itself, the selected version is the higher version v0.22.1.

There's dramatic dependency change between v0.20.1 and v0.22.1.
https://github.com/census-instrumentation/opencensus-go/commit/ed3a3f0bf00d34af1ca7056123dae29672ca3b1a#diff-37aff102a57d3d7b797f152915a6dc16
In v0.20.1, it has several exporter related packages, like zipkin-go which uses mux v1.6.2: https://github.com/openzipkin/zipkin-go/blob/v0.1.6/go.mod#L14

Before Go modules decides which version to select, it needs to construct the complete build list first. That's to say, it appends dependencies of both v0.20.1 and v0.22.1 even if v0.20.1 is not selected eventually.

To summarize, if ES client prunes `google.golang.org/api v0.3.1` from its go.mod, there's no version bumping for `github.com/gorilla/mux`, `github.com/prometheus/client_golang`, `github.com/prometheus/common` etc.

**Comment by harudark** [comment]
> @harudark Should we close MRs #2426, #2427 and #2442 then?

I'll fix issues mentioned in code review and then close those MRs.

File: go.sum (hunk 4 above)
[the remark this answers is not in the document]
Comment (muffix) [reply go.sum]: Cool, happy to merge this once that's done (and rebased on the latest master). Thanks again for the contribution. 🙌

**Comment by langerma** [comment]
any news on that?

**Comment by johnewing1** [comment]
@svangner @harudark 
Have you had a chance to make the last few changes to this ?

File: go.mod (hunk 1 above)
Comment (johnewing1) [reply go.mod:47]: Looks like this is now fixed in the upstream elastic client
https://github.com/olivere/elastic/commit/5d20c799557a7637d81468010fc1246cef0b8ec6

Comment (johnewing1) [reply go.mod]: Updated the go.mod and go.sum based on the newest version and it's no longer pulling in the unneeded deps
